test/server/core/gameSteps/regroupWindow/RegroupPhase.spec.ts

```diff
@@ -0,0 +1,144 @@
+                    expect(context.player2.hand).toEqual(oldHandPlayer2);
+
+                    // exhausted units need to be exhausted until the ready card phase
+                    expect(context.allianceXwing.exhausted).toBe(true);
+                    expect(context.wampa.exhausted).toBe(true);
+                    expect(context.tielnFighter.exhausted).toBe(true);
+
+                    // Resource a Card
+                    // Player 1 Resources a card and Player 2 doesn't
+                    context.player1.clickCard('wroshyr-tree-tender');
+                    // this is the index of wroshyr-tree-tender
+                    oldResourcesPlayer1.push(oldHandPlayer1[1]);
+                    context.player1.clickPrompt('Done');
+                    context.player2.clickPrompt('Done');
+
+                    // check resources
+                    expect(context.player1.resources.length).toBe(5);
+                    expect(context.player2.hand).toEqual(oldHandPlayer2);
+                    expect(context.player1.resources).toEqual(oldResourcesPlayer1);
+                    expect(context.player2.resources).toEqual(oldResourcesPlayer2);
+
+                    // ready card phase
+                    expect(context.allianceXwing.exhausted).toBe(false);
+                    expect(context.wampa.exhausted).toBe(false);
+                    expect(context.tielnFighter.exhausted).toBe(false);
+                    expect(context.player2).toBeActivePlayer();
+
+                    // Case 3 player 2 can only draw 1 card and receives 3 damage to base;
+                    context.player2.passAction();
+                    context.player1.claimInitiative();
+
+                    // Draw cards
+                    expect(context.player1.hand.length).toBe(6);
+                    expect(context.player2.hand.length).toBe(4);
+
+                    // Resources
+                    context.player2.clickPrompt('Done');
+                    context.player1.clickPrompt('Done');
+
+                    // check board state
+                    expect(context.player2.deck.length).toBe(0);
+                    expect(context.p2Base.damage).toBe(3);
+                }
+            );
+
+            it('should end all "for this phase" abilities',
+                function () {
+                    contextRef.setupTest({
+                        phase: 'action',
+                        player1: {
+                            hand: ['attack-pattern-delta'],
+                            groundArena: ['ardent-sympathizer', 'scout-bike-pursuer'],
+                            spaceArena: ['alliance-xwing'],
+                        },
+                        player2: {
+                            groundArena: ['wampa'],
+                        }
+                    });
+
+                    const { context } = contextRef;
+
+                    context.player1.clickCard(context.attackPatternDelta);
```

Comment:
This is great! Couple additional things would be good:

- Can you add break up this block of code with some comments explaining what the clicks are doing, it's a little hard to follow
- It would also be good to explicitly test the timing on the defeat. General Krell is fantastic for that, could you add him for player 1 and confirm that when his ability prompt comes up that we are still in the action phase

Reply:
Great idea! I'll add him to the tests